xds/src/main/java/io/grpc/xds/sds/TlsCertificateSecretVolumeSecretProvider.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.sds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ *  Implementation of a file based secret provider.
+ */
+final class TlsCertificateSecretVolumeSecretProvider
+    implements SecretProvider<TlsCertificateStore> {
+
+  private static final Logger logger = Logger
+      .getLogger(TlsCertificateSecretVolumeSecretProvider.class.getName());
+  public static final String PEM = ".pem";
+  public static final String CRT = ".crt";
+
+  private String path;
+
+  // for now mark it unused
+  @SuppressWarnings("unused")
+  private String name;
+
+  TlsCertificateSecretVolumeSecretProvider(String path, String name) {
+    this.path = path;
+    this.name = name;
+  }
+
+  @Override
+  public void addListener(Runnable listener, Executor executor) {
+    checkNotNull(listener, "Runnable was null.");
+    checkNotNull(executor, "Executor was null.");
+    try {
+      executor.execute(listener);
+    } catch (RuntimeException e) {
+      // ListenableFuture's contract is that it will not throw unchecked exceptions, so log the bad
+      // runnable and/or executor and swallow it.
+      logger.log(
+          Level.SEVERE,
+          "RuntimeException while executing runnable " + listener + " with executor " + executor,
+          e);
+    }
+  }
+
+  @Override
+  public boolean cancel(boolean mayInterruptIfRunning) {
+    return false;
+  }
+
+  @Override
+  public boolean isCancelled() {
+    return false;
+  }
+
+  @Override
+  public boolean isDone() {
+    return true;
+  }
+
+  /**
+   * Gets the current contents of the private key and cert file Assume the key has .pem extension
+   * and cert has .crt extension (needs to match how Citadel mounts secrets)
+   */
+  @Override
+  public TlsCertificateStore get() throws InterruptedException, ExecutionException {
```

Comment:
If you call `get()` twice, it will return different instances, that's not Java Future class should behave.

Reply:
Looking up the Javadoc for [`Future.get()`](https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/Future.html#get()) it doesn't promise idempotency there. I did see one references about that assumption [here](https://guava.dev/releases/6.0/api/docs/com/google/common/util/concurrent/Futures.html#compose(java.util.concurrent.Future,%20com.google.common.base.Function)) .

I do understand the concern though. The `get()` in this case is supposed to return the current certificate content as loaded from the store so that functionality needs to be there. I can transfer the logic to TlsCertificateStore and make sure the same instance is returned by `get()` if that's acceptable. Or have subclasses for TlsCertificateStore

Reply:
SecretProvider.get is no longer that of Future.get. The ListenableFuture logic now is now inside individual SecretProvider. PTAL